dascore/proc/coords.py

```diff
@@ -194,3 +197,96 @@ def drop_coords(self: PatchType, *coords: str | Collection[str]) -> PatchType:
         raise ParameterError(msg)
     new_coord, data = self.coords.drop_coords(*coords, array=self.data)
     return self.new(coords=new_coord, dims=new_coord.dims, data=data)
+
+
+@patch_function()
+def coords_from_df(
+    self: PatchType, dataframe, units=None, extrapolate=True
+) -> PatchType:
+    """
+    Add coordinates to a patch using a dataframe.
+
+    One of the column names in the dataframe should map to one of the patch.dims
+
+    Will either add new coordinates, or update existing ones.
+
+    Parameters
+    ----------
+    dataframe : pandas dataframe
+        Table with a column matching in title to one of patch.dims along with other
+        coordinates to associate with dimension. Example one column matching distance
+        axis and then latitude and longitude attached to the distances.
+    units : dict, optional
+        dictionary mapping added coordinates to their units
+    extrapolate : str, optional
+        Extrapolate outside of provided range in dataframe to
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import dascore as dc
+    >>> import pandas as pd
+    >>> # get example patch and coordinates
+    >>> pa = dc.get_example_patch()
+    >>> path = fetch("brady_hs_DAS_DTS_coords.csv")
+    >>> coord_table = pd.read_csv(path)
+    >>> # get distance axis
+    >>> dist = coords.get_array('distance')
+    >>> # interpolate to get corresponding new coordinates for
+    >>> # comparison
+    >>> X_0 = np.interp(
+                    axis_coords,
+                    pd.to_numeric(coord_table["distance"]),
+                    pd.to_numeric(coord_table["X]),
+                    left=float("nan"),
+                    right=float("nan"),
+                )
+    >>> # attach new coordinates to patch
+    >>> patch = patch.coords_from_df(pa, df1, units=units)
+    >>> # retrieve coordinates and compare to expected
+    >>> coords = patch.coords
+    >>> X = coords.get_array('X')
+    >>> assert np.all(X == X_0)
+    """
+    # match dataframe headings to dims
+    axis_to_update = set(self.dims) & set(dataframe.columns)
+
+    if len(axis_to_update) != 1:
+        raise Exception("Exactly one column has to match with an existing dimension")
+
+    # Get coordinates of axis being updated
+    axis_to_update = next(iter(axis_to_update))
+    coords = self.coords
+    axis_coords = coords.get_array(axis_to_update)
+
+    # make a dictionary from coordinates("(axis, coordinate array)") as input to
+    # update_coords
+    # coordinate array is an interpolation to match existing coords being updated
+    new_coords = {}
+
+    for coord in dataframe.columns:
```

Comment:
This is probably more clear:

```python 
for coord in set(dataframe.columns) - {axis_to_update}:
    if extrapolate: 
        ...
    else:
        ...
```

Also, `is True` is usually not the right way to check truthiness. From pep8:

![image](https://github.com/DASDAE/dascore/assets/11671536/7105e764-6118-4088-b6c3-1f57b8beee34)